qiskit/providers/models/backendconfiguration.py

```diff
@@ -243,53 +263,73 @@ def __init__(self, backend_name, backend_version, n_qubits, basis_gates,
-                 discriminators, **kwargs):
-
+    def __init__(self,
+                 backend_name: str,
+                 backend_version: str,
+                 n_qubits: int,
+                 basis_gates: List[str],
+                 gates: GateConfig,
+                 local: bool,
+                 simulator: bool,
+                 conditional: bool,
+                 open_pulse: bool,
+                 memory: bool,
+                 max_shots: int,
+                 n_uchannels: int,
+                 u_channel_lo: List[UchannelLO],
+                 meas_levels: List[int],
+                 qubit_lo_range: List[List[float]],
+                 meas_lo_range: List[List[float]],
+                 dt: float,
+                 dtm: float,
+                 rep_times: List[float],
+                 meas_kernels: List[str],
+                 discriminators: List[str],
+                 **kwargs):
+        """
+        Initialize a backend configuration that contains all the extra configuration that is made
+        available for OpenPulse backends.
+
+        Args:
+            backend_name: backend name.
+            backend_version: backend version in the form X.Y.Z.
+            n_qubits: number of qubits.
+            basis_gates: list of basis gates names on the backend.
+            gates: list of basis gates on the backend.
+            local: backend is local or remote.
+            simulator: backend is a simulator.
+            conditional: backend supports conditional operations.
+            open_pulse: backend supports open pulse.
+            memory: backend supports memory.
+            max_shots: maximum number of shots supported.
+            n_uchannels: Number of u-channels.
+            u_channel_lo: U-channel relationship on device los.
+            meas_levels: Supported measurement levels.
+            qubit_lo_range: Qubit lo ranges for each qubit with form (min, max) in GHz.
+            meas_lo_range: Measurement lo ranges for each qubit with form (min, max) in GHz.
+            dt: Qubit drive channel timestep in nanoseconds.
+            dtm: Measurement drive channel timestep in nanoseconds.
+            rep_times: Supported repetition times for device in microseconds.
+            meas_kernels: Supported measurement kernels.
+            discriminators: Supported discriminators.
+            **kwargs: Optional fields.
+        """
         self.n_uchannels = n_uchannels
         self.u_channel_lo = u_channel_lo
         self.meas_levels = meas_levels
         self.qubit_lo_range = qubit_lo_range
         self.meas_lo_range = meas_lo_range
-        self.dt = dt  # pylint: disable=invalid-name
-        self.dtm = dtm
+        self._dt = dt  # pylint: disable=invalid-name
+        self._dtm = dtm
         self.rep_times = rep_times
         self.meas_kernels = meas_kernels
         self.discriminators = discriminators
+        if 'hamiltonian' in kwargs:
+            self._hamiltonian = kwargs.pop('hamiltonian')
```

Comment:
The way to have optional parameters is with `hamiltonian=None` in the `__init__` definition. I thinks that is clearer.